File: server/services/ebook.py
#!/usr/bin/env python3
import sys
import json
import os
import re
import argparse
from pathlib import Path

# Add user packages to path
import subprocess
result = subprocess.run(['python3', '-m', 'site', '--user-site'], capture_output=True, text=True)
user_site = result.stdout.strip()
if user_site and user_site not in sys.path:
    sys.path.insert(0, user_site)

try:
    import ebooklib
    from ebooklib import epub
    from mobi import extract as mobi_extract
except ImportError as e:
    print(json.dumps({"error": f"Missing dependency: {e}"}))
    sys.exit(1)

def clean_html(text):
    """Remove HTML tags and clean text"""
    if not text:
        return ""
    # Remove HTML tags
    text = re.sub(r'<[^>]+>', ' ', text)
    # Remove extra whitespace
    text = re.sub(r'\s+', ' ', text)
    return text.strip()

def process_epub(epub_path):
    """Extract text and metadata from epub file"""
    book = epub.read_epub(epub_path)
    
    metadata = {
        "title": None,
        "author": None,
        "description": None,
        "language": None,
        "publisher": None,
        "identifier": None
    }
    
    # Get metadata
    for item in book.get_metadata('DC', 'title'):
        if item and item[0]:
            metadata['title'] = item[0]
            
    for item in book.get_metadata('DC', 'creator'):
        if item and item[0]:
            metadata['author'] = item[0]
            
    for item in book.get_metadata('DC', 'description'):
        if item and item[0]:
            metadata['description'] = clean_html(str(item[0]))
            
    for item in book.get_metadata('DC', 'language'):
        if item and item[0]:
            metadata['language'] = item[0]
            
    for item in book.get_metadata('DC', 'publisher'):
        if item and item[0]:
            metadata['publisher'] = item[0]
    
    # Extract text
    full_text = ""
    for item in book.get_items():
        if item.get_type() == 9:  # DOCUMENT
            content = item.get_content()
            if isinstance(content, bytes):
                content = content.decode('utf-8', errors='ignore')
            clean_text = clean_html(content)
            if len(clean_text) > 50:
                full_text += clean_text + " "
    
    return {
        "text": full_text.strip(),
        "metadata": metadata,
        "word_count": len(full_text.split()),
        "char_count": len(full_text)
    }

def process_mobi(mobi_path):
    """Convert mobi to epub and extract"""
    # Convert mobi to epub
    epub_path = mobi_extract(mobi_path)
    return process_epub(epub_path)

def main():
    parser = argparse.ArgumentParser(description='Process ebook files')
    parser.add_argument('file', help='Path to ebook file')
    args = parser.parse_args()
    
    file_path = args.file
    
    if not os.path.exists(file_path):
        print(json.dumps({"error": f"File not found: {file_path}"}))
        sys.exit(1)
    
    ext = os.path.splitext(file_path)[1].lower()
    
    try:
        if ext == '.mobi':
            result = process_mobi(file_path)
        elif ext == '.epub':
            result = process_epub(file_path)
        else:
            print(json.dumps({"error": f"Unsupported format: {ext}"}))
            sys.exit(1)
        
        print(json.dumps(result))
        
    except Exception as e:
        print(json.dumps({"error": str(e)}))
        sys.exit(1)

if __name__ == '__main__':
    main()

File: server/services/ebook.js
import { exec } from 'child_process';
import { promisify } from 'util';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const execAsync = promisify(exec);
const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Convert mobi to epub using Python
async function convertMobiToEpub(mobiPath) {
  const { extract } = await import('mobi');
  const outputDir = path.dirname(mobiPath);
  try {
    const epubPath = await extract(mobiPath);
    return epubPath;
  } catch (e) {
    console.error('Mobi conversion error:', e.message);
    throw e;
  }
}

// Extract text from epub
async function extractEpubText(epubPath) {
  const { read_epub } = await import('ebooklib');
  const book = await read_epub(epubPath);
  
  let fullText = '';
  let metadata = {
    title: null,
    author: null,
    description: null
  };
  
  // Get metadata
  try {
    const title = book.getMetadata('DC', 'title');
    if (title.length > 0) metadata.title = title[0][0];
  } catch (e) {}
  
  try {
    const author = book.getMetadata('DC', 'creator');
    if (author.length > 0) metadata.author = author[0][0];
  } catch (e) {}
  
  try {
    const desc = book.getMetadata('DC', 'description');
    if (desc.length > 0) metadata.description = desc[0][0];
  } catch (e) {}
  
  // Extract text from all items
  for (const item of book.getItems()) {
    if (item.getType() === 9) { // DOCUMENT
      const content = item.get_content();
      if (Buffer.isBuffer(content)) {
        const text = content.toString('utf-8');
        // Strip HTML tags
        const cleanText = text.replace(/<[^>]+>/g, ' ').replace(/\s+/g, ' ').trim();
        if (cleanText.length > 50) {
          fullText += cleanText + ' ';
        }
      }
    }
  }
  
  return { text: fullText.trim(), metadata };
}

// Main function to process ebook
export async function processEbook(filePath) {
  const ext = path.extname(filePath).toLowerCase();
  let epubPath = filePath;
  let originalFile = filePath;
  
  // Convert mobi to epub if needed
  if (ext === '.mobi') {
    console.log('Converting mobi to epub...');
    epubPath = await convertMobiToEpub(filePath);
    console.log('Converted to:', epubPath);
  }
  
  if (ext !== '.epub') {
    throw new Error('Unsupported file format. Only epub and mobi are supported.');
  }
  
  // Extract text
  console.log('Extracting text from epub...');
  const { text, metadata } = await extractEpubText(epubPath);
  
  console.log('Extracted text length:', text.length);
  console.log('Metadata:', metadata);
  
  return {
    text,
    metadata,
    wordCount: text.split(/\s+/).length,
    charCount: text.length
  };
}
